### PR TITLE
make sub of timedelta work

### LINKED ISSUE #758: No overload variant of "__sub__" of "Series" matches argument type "Timedelta"
**Describe the bug**
After upgrading to pandas 2.0 started getting errors as described in the title

**To Reproduce**
```python
import pandas as pd

df = pd.read_csv("afile.csv", parse_dates=["foo"])
df["foo"] = df["foo"] - pd.Timedelta("1D")
```

Using mypy:

```sh
pandas-stubs-mre.py:4: error: No overload variant of "__sub__" of "Series" matches argument type "Timedelta"  [operator]
pandas-stubs-mre.py:4: note: Possible overload variants:
pandas-stubs-mre.py:4: note:     def __sub__(self, Timestamp | datetime | TimestampSeries, /) -> TimedeltaSeries
pandas-stubs-mre.py:4: note:     def __sub__(self, complex | ExtensionArray | ndarray[Any, Any] | dict[str, ndarray[Any, Any]] | list[Any] | tuple[Any, ...] | Index | Series[Any], /) -> Series[Any]
Found 1 error in 1 file (checked 1 source file)
```

**Please complete the following information:**
 - OS: Discovered on both Windows / Linux
 - python version: 3.11.0
 - version of type checker: mypy 1.4.1 (compiled: yes)
 - version of installed `pandas-stubs`: 2.0.2.230605

**Additional context**
The issue does not happen with older versions of pandas-stubs - 1.5.3.230321 issues no complaint. I haven't done a full bisect but guess this may go back to https://github.com/pandas-dev/pandas-stubs/pull/616 ?

### PR DESCRIPTION
**Pull request by Dr-Irv**
- [x] Closes #758 
- [x] Tests added: Please use `assert_type()` to assert the type of any return value
   - `test_timefuncs.py:test_timedelta64_and_arithmatic_operator()` - added tests there

### CODE REVIEW

**Review by twoertwein — COMMENTED** [review]
Looks good to me but there might be a numpy-pyright error

**Comment by Dr-Irv** [comment]
> Looks good to me but there might be a numpy-pyright error

Actually, it was an issue with pyright 1.1.320.  Needed to test differently on mypy vs. pyright, which I figured out how to do.  Take another look

File: tests/test_indexes.py [hunk 1]
```diff
@@ -24,6 +24,15 @@
...
 if TYPE_CHECKING:
+    MYPY_CHECKING: bool = True
+    if MYPY_CHECKING:
+        from typing import Any
+
+        from typing_extensions import TypeAlias
+
+        SupportsRichComparison: TypeAlias = Any
+    else:
+        from _typeshed import SupportsRichComparison  # noqa: F401
```
Comment (twoertwein) [inline tests/test_indexes.py:35]: Do you mind referencing an issue or describing the discrepancy? Is it just that mypy has an older version of typestubs?

Comment (Dr-Irv) [reply tests/test_indexes.py:37]: Added comments in https://github.com/pandas-dev/pandas-stubs/pull/759/commits/5310eb682334877ae00a06b80082fc80f7467aee

**Comment by twoertwein** [comment]
Thanks @Dr-Irv !

File: tests/test_indexes.py (hunk 1 above)
Comment (Dr-Irv) [reply tests/test_indexes.py:37]: > Does this also work for `poe mypy --mypy_nightly`?

Looks like it did!

We may want to use this pattern (`MYPY_CHECKING`) elsewhere when `mypy` and `pyright` make different inferences.  Can't put it in the stubs themselves - that would require users to define `MYPY_CHECKING` as `false` if using pyright, but we can use it in our tests.

**Pull request merged**